Move servicebus to sdk

mitchdenny opened the pull request:
Moves code from ```servicebus/data-plane/``` to ```sdk/servicebus```.

mitchdenny commented:
@JonathanGiles for input on the Java side.
@weshaggard for input on the approach I'm taking with the streamlined pipelines.

JonathanGiles commented:
More generally, I am not clear how track two service bus will be expected to co-exist in this directory structure: https://github.com/mitchdenny/azure-sdk-for-java/tree/move-servicebus-to-sdk/sdk/servicebus @weshaggard Can you speak to the plan here?

mitchdenny commented:
Hrm - good point. Seems that there are a few options here (don't think this is exhaustive):

1. Co-exist with different library names.
2. Do servicing for the track 1 version on a branch.
3. Converge the code into a single library (and when appropriate deprecate the old classes/interfaces).

This kind of problem will likely come up again and again so it would be good to have a general strategy we apply.

weshaggard commented:
The guidelines I've currently outlined in https://github.com/Azure/azure-sdk/blob/master/docs/engineering-system/repo-structure.md.

- **`<package name>`** - Should be the name of the shipping package, or an abbreviation that distinguishes the given shipping artifact for the given service. The key is that there is source for only one shipping package in this folder.
    - If there are multiple packages with the same name but different scope/org/groupid/SxS-version/etc then put them each in their own folder under the service name directory with names that match whatever distinguishes them from each other.
    - If there are long file path issues then you can use an abbreviation of the package name to help avoid issues. For example Java tools require every part of the package/namespace to be a separate folder which can make the paths extra long and thus may need to abbreviate the package name to help alleviate the impact of long file paths on windows.
    - If there are other assets that aren't shipping packages, such as shared libaries/source code or tools, they can also go in a folder under the service name directory.


So based on those it boils down to whether or not there they will share the same artifactid and/or ship SxS. @JonathanGiles correct me if I'm wrong but I believe for the Java track 2 libraries they are going to be new libraries SxS with only the groupId changed from `com.microsoft.azure` to `com.azure`? If that is the case then I'll suggest the following convention:

- track 1: package folders should be named `microsoft-<artifactid>`
- track 2: package folders should be named `<artifactid>`

Does that sound reasonable to you @JonathanGiles?

mitchdenny commented:
Thanks for that @weshaggard! That sounds like a reasonable plan to me and it'll allow us to continue servicing off the master branch for the older package versions.

JonathanGiles commented:
I know I've raised this concern before. I'll outline it again now because there is a little more concrete detail thanks to this PR. For reference, here is the directory structure of the move: https://github.com/mitchdenny/azure-sdk-for-java/tree/move-servicebus-to-sdk/sdk/servicebus

The concern I have is that there are two pom.xml files for this track 1 move, within the /sdk/servicebus directory. There is a /sdk/servicebus/pom.xml and a /sdk/servicebus/azure-servicebus/pom.xml. The latter /sdk/servicebus/azure-servicebus/pom.xml refers to the /sdj/servicebus/pom.xml as its parent pom.

Introducing track two will mean we add additional directories under /sdk/servicebus, presumably in, e.g. servicebus. So that means we will have /sdk/servicebus/servicebus for track two, and /sdk/servicebus/azure-servicebus. To me, this is not a clear enough difference.

Additionally, this approach requires reworking of the pom files for each move, assuming we keep the 'root' /sdk/servicebus/* files - these are all explicitly track one files.

If it were me, I would like to have /sdk/servicebus/data-plane (where all files from this move would go into), and /sdk/servicebus/client (where track two will go). This makes moving vastly simpler, and in my opinion makes identifying modules and which are relevant to a users interest far simpler.

mitchdenny commented:
I think that would make the migration easier for the track 1 libraries. I'd also like to think over the horizon a little bit here and I wonder how we would approach revving these libraries if we decide that the external library types that we are exposing on our public interfaces aren't all the rage anymore and we want to really shake up our API - do we end up creating another folder under ```sdk/[service]```.

Or - will we be evolving the API in place whilst managing deprecation etc?

weshaggard commented:
>If it were me, I would like to have /sdk/servicebus/data-plane (where all files from this move would go into), and /sdk/servicebus/client (where track two will go). This makes moving vastly simpler, and in my opinion makes identifying modules and which are relevant to a users interest far simpler.

Thanks for the feedback @JonathanGiles I know you are still a little uneasy about having these mixed but I think I'd still like to continue with this path for now. The hope is that each service directory will have a ReadMe.md that has information about each of the directories/packages and will help guide folks to better understanding what they are and which ones they should pay attention to, I already chatted with @kaerm about this adding that information to the readmes. 

If in the end we figure out this causes too many issued then we will consider other options, one that I'm thinking about is having another root level service folder like "servicebus-old"  (not sold on the name yet) to hold the track 1 packages and then use "servicebus" for the track 2 and management libraries which are still actively developed.

mitchdenny commented:
OK - thanks all, I'd like to move forward for now with this then - @JonathanGiles do you know who I need to ping specifically around the servicebus folder move here? Or can you give this one a tick of approval?

weshaggard commented on eng/pipelines/templates/variables/globals.yml:
```diff
@@ -0,0 +1,3 @@
+variables:
+  DefaultOptions: '--batch-mode -Dmaven.wagon.http.pool=false'
+  LoggingOptions: '-Dorg.slf4j.simpleLogger.defaultLogLevel=error -Dorg.slf4j.simpleLogger.log.org.apache.maven.cli.transfer.Slf4jMavenTransferListener=warn'
```
I assume you also want to disable CG.

mitchdenny replied:
Fixed.

weshaggard commented on sdk/servicebus/.project:
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Any idea why these show up as adds vs moves? my guess is that for some reason git didn't delete the old one, most likely becaus eof the "." in the name caused it to be filtered.

mitchdenny replied:
Not sure - I actually did a clean clone to do the moves and did force ```git add . -f``` after I had made the tweaks just to make sure I got a clean move. I think it is some kind of Git quirk.

weshaggard replied:
Ok just make sure the old directory no longer exists after your change.

weshaggard commented:
@mssfang ported this code to the repo. @mssfang are you the owner of the servicebus java library? If so can you please take a look at this change and ensure it looks reasonable.

JonathanGiles commented:
There is a new service bus pull request you might want to reconcile with. 

In terms of my concerns, my primary worry here is the extra work we will have to do when a track two library is added. It will break track one developers on that library, it will confuse people about directory structure, there will need to be additional refactoring for every client move, etc. Anyway, I have no additional feedback beyond these concerns, so feel free to merge as desired.

mssfang commented:
This PR includes java complied ***.class files. Please make sure to exclude them before merging to master.
Additionally, FYI, the new PR from the service team merged into the master recently but no more incoming PR for the next few days.   Seems there are no Java and Maven code changes, it looks fine to me.

mitchdenny commented:
Thanks folks - I'll fix the PR up.

weshaggard commented:
@mitchdenny for the libraries you are moving that we know are track 1, can you also add the `microsoft-` prefix to help avoid another move for them when their track 2 library comes in?

mitchdenny commented:
Dunno what happened there - didn't mean to close the PR. I've updated the PR against the changes in master. I've renamed ```servicebus``` to ```microsoft-servicebus``` to make room for the future track 2 version of the library.

mitchdenny commented:
Merging this now.

JonathanGiles commented:
This is still bizarre to me. We now have `microsoft-servicebus` for track one, and the artifactId of that product is 'azure-servicebus' (which will also likely be what we name our track two library).

weshaggard commented:
This was supposed to be `microsoft-azure-servicebus` (fixed in https://github.com/Azure/azure-sdk-for-java/pull/3911) and I expect our track 2 will be `azure-servicebus`. Regardless we are going to have to explain to customers which one is which and we will do it in the readme in the service folder and most likely some other public facing documentation when we ship.
